plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/rest/TrinoIcebergRestCatalogFactory.java

```diff
@@ -0,0 +1,92 @@
+
+import com.google.common.collect.ImmutableMap;
+import io.airlift.log.Logger;
+import io.trino.hdfs.ConfigurationUtils;
+import io.trino.plugin.base.CatalogName;
+import io.trino.plugin.hive.NodeVersion;
+import io.trino.plugin.iceberg.IcebergConfig;
+import io.trino.plugin.iceberg.catalog.TrinoCatalog;
+import io.trino.plugin.iceberg.catalog.TrinoCatalogFactory;
+import io.trino.spi.security.ConnectorIdentity;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.rest.RESTSessionCatalog;
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+
+import javax.inject.Inject;
+
+import java.net.URI;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class TrinoIcebergRestCatalogFactory
+        implements TrinoCatalogFactory
+{
+    private static final Logger log = Logger.get(TrinoIcebergRestCatalogFactory.class);
+
+    private final CatalogName catalogName;
+    private final String trinoVersion;
+    private final URI serverUri;
+    private final IcebergRestCatalogConfig.Security security;
+    private final Optional<String> credential;
+    private final Optional<String> token;
+    private final boolean isUniqueTableLocation;
+
+    private volatile RESTSessionCatalog icebergCatalog;
+
+    @Inject
+    public TrinoIcebergRestCatalogFactory(
+            CatalogName catalogName,
+            IcebergRestCatalogConfig restConfig,
+            IcebergConfig icebergConfig,
+            NodeVersion nodeVersion)
+    {
+        this.catalogName = requireNonNull(catalogName, "catalogName is null");
+        this.trinoVersion = requireNonNull(nodeVersion, "nodeVersion is null").toString();
+        requireNonNull(restConfig, "restConfig is null");
+        this.serverUri = restConfig.getBaseUri();
+        this.credential = restConfig.getCredential();
+        this.token = restConfig.getToken();
+        this.security = restConfig.getSecurity();
+        requireNonNull(icebergConfig, "icebergConfig is null");
+        this.isUniqueTableLocation = icebergConfig.isUniqueTableLocation();
+    }
+
+    @Override
+    public synchronized TrinoCatalog create(ConnectorIdentity identity)
+    {
+        if (icebergCatalog == null) {
+            ImmutableMap.Builder<String, String> properties = ImmutableMap.builder();
+            properties.put(CatalogProperties.URI, serverUri.toString());
+            properties.put("trino-version", trinoVersion);
+            if (security == IcebergRestCatalogConfig.Security.OAUTH2) {
+                checkArgument(credential.isPresent() ^ token.isPresent(), "OAuth2 requires either a credential or token");
+                credential.ifPresent(value -> properties.put(OAuth2Properties.CREDENTIAL, value));
+                token.ifPresent(value -> properties.put(OAuth2Properties.TOKEN, value));
+            }
+
+            RESTSessionCatalog icebergCatalogInstance = new RESTSessionCatalog();
```

Comment:
Any reason why this cannot be done eagerly within  the constructor?
what does this laziness give us?

Reply:
The RESTCatalog will initialize with the config route call and OAuth2 exchanges.  I just wasn't clear if there is a preference for doing that initialization eagerly or lazily.  I'm fine either way if there's a precedent.  Just let me know what you want.

Reply:
The only comparable example I see is the Hive metastore, which is similarly lazy (via `memoizeMetastore`).  I think I initially took that as the example.

Reply:
Is the `RESTSessionCatalog` init does a network call,   it indeed needs to be lazy. We cannot assume some other service is up when Trino is starting up (for example: Trino and the metastore / catalog services can be restarted concurrently).


I am fine doing this lazy but please have a code comment why doing so